squid api squid next iteration revision #2

billbsing opened the pull request:
## Description

This will probably break everything! 

I thought i would add a possible re-design which is slightly different to the current Squid API.

I've replaced DDO with meta data, as the end user calling the library just needs to know about the metadata, and the squid library can create the DDO and include the metadata.

I'm also looking at this with a Python point of view. So with the squid-py the calls will have to be underscore_case instead of camelCase

H34D committed:
added accounts, separated service requests and orders, added pseudocode for ocean object, moved all the objects accounts, assets, orders, service agreements under ocean

H34D committed:
reflect hierarchy more in table display, added DID object, smaller changes

H34D committed:
update exmaple

H34D committed:
removed old squid api

H34D committed:
fix TOC

H34D committed:
re-added getAccounts

billbsing committed:
squid api

aaitor commented on doc/development/squid.md:
```diff
@@ -71,182 +85,218 @@ Parameters:
 
 #### Functions
 
+* **getAccounts** - Returns all avalible accounts loaded via a wallet, or by Web3.
```
IMO this function is not a Must Have. 
Not having an accounts registry, it's difficult to retrieve all the accounts interacting with the system. I'm not sure if it's necessary or recommended either.

H34D replied:
i think it is, here is an excerpt from my PR:

> re added getAccounts because we need them. We don't have an actor registry but we still have accounts (that are coming from the blockchain node). We need access to those informations. Maybe accounts is the wrong term. Shall we call it wallets maybe? Because thats how we use it right now. We can topple up wallets (requestTokens), we can list wallets (getAccounts).

billbsing replied:
Certainly agree it's a must have. Missed that one, I was going with the Python style of removing the 'get' from the property name, so ```getAccounts``` became ```accounts```

To make things easier we can just return the account address, similar to Web3

billbsing replied:
I've been stuck on wallets for a few weeks, trying to solve this one. Possibly copy [Web3.js](https://web3js.readthedocs.io/en/1.0/web3-eth-accounts.html)?

Where we load/create a wallet using a Wallet class, and then only pass the account address from the wallet to the squid library.

So the ```getAccounts``` in the squid library  is just a helper for Web3.getAccounts()

aaitor replied:
For clarify myself, is the expected behavior of this function, to get a list in your local node of all the existing accounts? Like the Parity Wallet client is doing?
For example, in squid-js will load this info from Metamask, in squid-py from the python wallet (ie. pywallet)

aaitor commented on doc/development/squid.md:
````diff
@@ -71,182 +85,218 @@ Parameters:
 
+* **searchAssets** - SYNC. Given a search query, returns a list of the DDO's matching with that query
+```
+array[assetDID]= ocean.searchAssets(searchQuery)
````
I think the return object should be an array of Asset's (Metadata)

billbsing replied:
I'm just being doing embedded stuff and I'm always trying to reduce memory usage. 
So yes we should be returning objects in the searches.

aaitor commented on doc/development/squid.md:
````diff
@@ -71,182 +85,218 @@ Parameters:
+```
 
-### Decentralized Identifiers (DID)
+* **getTrader** - ASYNC. Given an account address create a trader object
````
SYNC?

aaitor commented on doc/development/squid.md:
````diff
@@ -71,182 +85,218 @@ Parameters:
+```
 
-#### Functions
+* **searchOrders** - SYNC. Return a list of orders by search query. **Nice to Have**.
````
Orders are stored on-chain so the search could be made on top of the emitted events. We could check the events emitted during the ServiceAgreement process and expose the indexed attributes added to the event as search parameters.

billbsing replied:
I was thinking of this method to search the block chain for orders based on any account, including any new orders from the block chain order events.

Will we need to provide the squid user the ability to see past orders or other orders from other accounts?

aaitor replied:
IMO a user using the frontend application would require the past orders related to him. 
So I guess this function should search on events emitted by user address.

aaitor commented on doc/development/squid.md:
```diff
@@ -71,182 +85,218 @@ Parameters:
+### Asset 
 
-* **register** - ASYNC. High-level method publishing the metadata off-chain and registering the Service Agreement on-chain. It orchestrate the `publishMetadata` and `publishServiceAgreement` methods.
+* **getDDO** - ASYNC. Return the created DDO used by this asset.
```
this could be SYNC. getDDO should return the DDO & Metadata from OceanDB

billbsing replied:
I've corrected the non block chain/async items to SYNC

aaitor commented on doc/development/squid.md:
````diff
@@ -71,182 +85,218 @@ Parameters:
 ```
 
-* **publishMetadata** - SYNC. Given an asset DDO, register this DDO off-chain. It returns a boolean with the result of the operation.
+* **getDID** - ASYNC. Return the DID used by this asset.
````
I think this one is SYNC too

aaitor commented on doc/development/squid.md:
````diff
@@ -71,182 +85,218 @@ Parameters:
 ```
-
-* **updateMetadata** - SYNC. Given an asset DDO, update the DDO off-chain. This method replace the complete existing DDO by the DDO provided. It returns a boolean with the result of the operation.
+* **getId** - ASYNC. Return the Id used by this asset.
````
it's not the same than getDID?

aaitor commented on doc/development/squid.md:
```diff
@@ -71,182 +85,218 @@ Parameters:
+
+##### Functions
+
+* **getId** - ASYNC. Return the Id used by this order.
```
I think is SYNC

aaitor reviewed (commented):
Looks a good review. I added a few comments to check. Would be good if @ssallam and @H34D can take a look at it too.

H34D committed:
Merge remote-tracking branch 'origin/feature/squid_next_iteration' into feature/squid-next-gen-rev-2

# Conflicts:
#	doc/development/squid.md

H34D committed:
added function topics back to TOC
added sample for getting an ocean object in js right next to the python one
re added the interface definition of the ocean object
added ocean.xxx to all function call to emphasize that they are sub objects of the ocean object
moved registerAsset back to asset because it is asset related

H34D commented:
- encapsulate everything in one single object so that you only have to import the `ocean` object nothing else from squid
- placed all functionality (from ocean) to the ocean object. This gives a nice hierarchy like
   * Ocean
      * Accounts
      * Orders
      * ServiceAgreements
         * SecretStore
         * ...
      * ...

H34D commented:
> * encapsulate everything in one single object so that you only have to import the `ocean` object nothing else from squid

the reason for this: loading contracts + getting an instance of them on the blockchain is async and a relatively complex task. I do this if i get an instance of the ocean object and pass that instance down to all the children. This leads to faster load times and a smaller interface for the consumer if squid. 

instead of doing 
```javascript
const ocean = await Ocean.getInstance()
ocean.asset.registerAsset()
```
you would need to instantiate that object again or add an additional getter to the ocean object which makes it even worse. 

```javascript
const ocean = await Ocean.getInstance()
ocean.getAsset().registerAsset()
```
or even
```javascript
const ocean = await Ocean.getInstance()
const asset = getAsset(ocean)
asset.registerAsset()
```

H34D commented:
please find my code of the ocean object here: 
https://github.com/oceanprotocol/squid-js/blob/feature/new-ocean-interface/src/ocean/Ocean.ts
it's not aligned with the changes from this PR ofc

H34D commented:
Also i added tuna https://github.com/oceanprotocol/tuna a repository used to showcase example usages of `squid`. Actually only used for the JavaScript version.

H34D committed:
Merge remote-tracking branch 'origin/master' into feature/squid-next-gen-rev-2

billbsing committed:
make the getAccounts return the same as the web3 getaccounts

billbsing committed:
search methods return an array of objects

billbsing commented:
Many thanks for the input/help, I really appreciate it.

I noticed in the ```squid-js```, there are public classes that has been defined in the ```Ocean``` class, in this example I'm going to use ```Asset```.
So currently we have something like this:

```javascript
class Ocean {
    ....
    public asset: Asset;
    
    public getInstance() {
        ocean.asset = new Asset(this)
    }
}
class Asset {
   ...
   public methodForAsset() {}
}
// call method

ocean.asset.methodForAsset()
```

Here the Asset class is used as a libray manager/grouping of asset based calls, plus setting up and speeding up the async calls, but we will also have other classes that are being used as an object to do something on a single asset with an id/DID.

So we will have two different types of classes, with the same name, ```Asset```.

So I'm not sure what we should do, move the methods up one level to the ```Ocean``` class, and keep the ocean.asset private or rename one or both types of ```Asset``` classs.

billbsing commented:
I try to use the naming convention as  ```getObjectName(id/ref)```, to return an existing object within the data structure. But the actual method ```getObjectName()``` may have to create the object and load the relevant data.

Since I think we should be going for multiple instances of the ```Ocean``` object, we need to pass the object or Web3 object to any sub class.

I think it's a bit too much to hope the library user can correctly do, especially if some classes may have different constructors

```javascript
asset = new Asset(ocean);
````
So I started off implementing ```createObjectName```, so that the ```Ocean``` class would be able to create the sub object and with the correct initial values for that object .

So I ended up with the following:

```javascript
const ocean = await Ocean.getInstance()
asset = ocean.createAsset()
await asset.register()
```

but to make it simpler I decided to place the registerAsset in the Ocean class more as a helper method,
so ...

```javascript
const ocean = await Ocean.getInstance()
asset = await ocean.registerAsset()

class Ocean { 
....
    public async registerAsset() {
        asset = new Asset(this);
        await asset.register();
        return asset;
   }
```

billbsing committed:
fix SYNC

H34D commented:
> So I'm not sure what we should do, move the methods up one level to the `Ocean` class, and keep the ocean.asset private or rename one or both types of `Asset` classs.

if we would move all the methods up to the ocean object the class will become huge and no separation of concern could be done. Also it only contains only proxy methods to call the actual method.

H34D commented:
> So we will have two different types of classes, with the same name, `Asset`.

i now understand what you mean in combination with your code from https://github.com/oceanprotocol/dev-ocean/pull/65#issuecomment-429183644

We need models for the data structures Asset, Order and so. In the current layout (of the ocean object) this will lead to naming collisions (i already have them in my implementation of `squid-js`). I have an Asset object (that is member of ocean and holds the methods to interact with Assets) and an Asset Model in the models folder that is an description of the Asset data structure. If we want to go for a domain driven design we should not the asset logic asset. 

I have something like this in mind:

```typescript
import { Ocean } from "@oceanprotocol/squid"

const ocean: Ocean = await Ocean.getInstance()
const asset = ocean.registry.createAsset()
console.log("AssetId that was just created", asset.id)
```

whilst the code inside of createAsset() could be something like this

```typescript
function createAsset(): Asset {
  return {
    id: "0x8247392"
  } as Asset
}
```

We would need to find a terminology for "doing stuff in code" and one for "how is my data described"

We could also do something like this:

```typescript
import { Ocean, Asset } from "@oceanprotocol/squid"

const ocean: Ocean = await Ocean.getInstance()
const asset: Asset = new Asset("0x0342464")
ocean.registry.register(asset)
console.log("AssetId that was just created", asset.id)
```

maybe this is even more clear because we expose all the objects

billbsing commented:
I would prefer the 
```javascript
const ocean: Ocean = await Ocean.getInstance()
const asset = ocean.registry.createAsset()
```
Since we can pass the a reference/handle to the web3 or ocean object to the Asset constructor.

I would like to keep the Ocean class hierarchy as flat as possible, since I do not think we have that many methods/properties, being exposed from one class.

An example would be the Web3.eth class, most of the web3 functionality is exposed from this one class, and so makes the access/interface simpler.

H34D commented:
> I would prefer the
> 
> ```js
> const ocean: Ocean = await Ocean.getInstance()
> const asset = ocean.registry.createAsset()
> ```
> 
> Since we can pass the a reference/handle to the web3 or ocean object to the Asset constructor.
> 
> I would like to keep the Ocean class hierarchy as flat as possible, since I do not think we have that many methods/properties, being exposed from one class.
> 
> An example would be the Web3.eth class, most of the web3 functionality is exposed from this one class, and so makes the access/interface simpler.

i agree. that means we need different names for models and functionality.

H34D committed:
add return values

H34D committed:
added priorities

H34D committed:
updated return values, restructure

H34D committed:
added trivial model implementations

Commit:
Filling metadata api requests

H34D committed:
re-added account again, remoddeled account, worken on order

H34D committed:
removed access declaration

Commit:
add python compat get_instance

billbsing commented:
I have a few points..
1. before an Asset is registered, do we create the object, or have the library create the object. So

```javascript
asset: Asset = ocean.createAsset()
// or
asset: Asset = Asset(ocean)
```
2. getBalance.getEther, getBalance.getOcean
 or 
 getEtherBalance, getOceanBalance, both do the same thing

3. When creating the Asset, or when calling the ```Asset.register``` method, we need to provide an Ethereum account owner

H34D commented:
> 1. before an Asset is registered, do we create the object, or have the library create the object. So
> 
> ```js
> asset: Asset = ocean.createAsset()
> // or
> asset: Asset = Asset(ocean)
> ```
> 
> 1. getBalance.getEther, getBalance.getOcean
>    or
>    getEtherBalance, getOceanBalance, both do the same thing
> 2. When creating the Asset, or when calling the `Asset.register` method, we need to provide an Ethereum account owner

i just stumbled about all of them during my implementation right now:

1.
What i did and actually like because it is lean, this is my implementation in JavaScript:
```javascript
const publisherAccount: Account = ocean.getAccounts()[0]
const asset: Asset = new Asset(name, description, price)
ocean.register(asset, publisherAccount)
```

2.
* `getEtherBalance` returns the Balance of Ether on that Etherem Account
* `getOceanBalance` returns the Balance of Ocean Tokens (ERC20) on that Ethereum Account
* `getBalance` returns the Balance Object containing Ether and Ocean Balance

3.
We could use the Account object like i did in 1. and just pass it to the register method

ttmc reviewed (commented):
It seems this doc is still changing a lot, so I'll just make a few general comments at this point:

- This doc should either say what Squid is or it should link to a doc that does.
- Is the goal of this doc to help a developer develop a version of Squid in another programming language? If so:
  - shouldn't that be stated up-front?
  - it should be clear, from this doc, what should be implemented, i.e. without having to refer to any of the existing implementations. For example, for each method, the list of inputs and outputs should be clear.
- Not all common programming languages have classes (e.g. C, Go).

billbsing commented:
OK, so maybe we should do the following:

1. Specify that at the moment the squid API is currently defined for an OO type language such as JavaScript, Java, Python e.t.c.
1. Try to exclude coding examples
1. Specify just the calling objects and methods that will actually perform Ocean work ( such as publish, register, getMetadata ).

 I think we should design this API document as a OO style , since moving from OO design to a non OO library is easier, by  just renaming each class method and passing the first variable as the object. 
So for example:
```c++
asset.register(account_id, ...);

// would be become in c
asset_register(&asset_struct, account_id);

```
Another possible point of view is that the Squid API should not be 'officially' defined by Ocean, since it's just a tool/library to make it easier to access the Ocean ecosystem such as the smart contracts and  Provider.

ttmc commented:
@billbsing You're right, it's probably fine to start out assuming an OO language for now. Code examples should be fine because they help make abstract ideas more concrete, and therefore easier to understand.

You also wrote:
> Another possible point of view is that the Squid API should not be 'officially' defined by Ocean, since it's just a tool/library to make it easier to access the Ocean ecosystem such as the smart contracts and Provider.

That's an interesting idea and could save a lot of hassle, but I'm new to the project and don't know the decision-making that has led to the current design. I guess Squid could just be a set of tools for doing "more complicated things" (i.e. not containing any simple wrappers of web3 or Provider APIs, since Ocean application developers could just use those directly).

aaitor commented:
@ttmc , @billbsing 
Squid is a L2 api built on top of core Ocean components. Squid is a facilitator or enabler, but it's not the only way to interact with Ocean. A part of the Keeper, the other components can be or not used.
As part of the job of facilitate the usage of Ocean, we provide higher level API's to users.
Because of that, Squid is a Ocean Protocol project defined by Ocean team, of course getting feedback from outside collaborators. 
The community can implement alternative approaches in different languages using the Squid spec or in a totally independent way.
But it's part of our responsibility to facilitate the usage of Ocean, and Squid is a key part of that.

H34D committed:
return string on register asset because we dont need it back once we registered

MarcusJones committed:
update Sync table to text. Formatting.

MarcusJones committed:
Merge branch 'feature/squid-next-gen-rev-2' of https://github.com/oceanprotocol/dev-ocean into feature/squid-next-gen-rev-2

Commit:
Add intro re Squid API and goal of this doc

Commit:
Regenerated TOC using gh-md-toc
Some heading levels had to be changed so it would work properly.

MarcusJones committed:
update

MarcusJones committed:
update TOC

MarcusJones committed:
Merge branch 'feature/squid-next-gen-rev-2' of https://github.com/oceanprotocol/dev-ocean into feature/squid-next-gen-rev-2

MarcusJones committed:
update Sync Asset class with table

MarcusJones committed:
remove Get Asset ID is redundant, since Asset ID = DID

MarcusJones committed:
update Finished sync of table. Removed deleted methods.

MarcusJones committed:
update TOC

MarcusJones committed:
update TOC

MarcusJones committed:
update TOC

MarcusJones committed:
fix TOC levels

MarcusJones committed:
fix Intro

H34D committed:
adapted changes from squid discussion, updated interface definition

H34D committed:
update documentation based on reality ;)

aaitor commented:
we should merge this PR, it's lasting too long. After merging we can continue refactoring the API.
Somebody disagree?

ttmc reviewed (approved):
I'm approving this. It seems to be in pretty good shape. If it has errors or inconsistencies (which it might), those can be fixed in future pull requests.
